### PR TITLE
Feature/replace to rspec

### PR DESCRIPTION
**Pull request by KensoDev**
Replace Minitest with Rspec.
I tested generating model and some other stuff, all working with rspec.
Main Vendored gem still works with Minitest.

### CODE REVIEW

**Comment by brandur** [comment]
@mfine @pedro Hey guys, any thoughts on this one? I've probably already covered this in depth, but I'm personally not the biggest fan of Minitest (although I will say that it works great for smaller projects). These are probably my top two problems with Minitest:

### Poor command tooling

To run the whole suite, you generally use Rake's `TestTask`; that part is okay. To run a single suite you generally use something like `ruby -Itest path/to/test` or `rake test TEST=path/to/test`; these are okay, but especially with the first `ruby` command, it's not great how you have to initialize a path to get `test_helper` or use `require_relative`.

The worst is the one-off test. Generally you use something like `rake -Itest path/to/test -n /test_pattern/`, but other types of specificity (like line numbers) aren't built in. You also have to know the huge gotcha that a spec like `it "tests for success"` internally gets converted to `test_for_success` and that you have to match on the latter with `-n`.

### Unsanitized backtraces

Minitest dumps big and ugly 50 line traces, including _everything_. RSpec will show you the lines that failed inside your app.

There are fixes for these things that normally ship with extension gems, but you kind of have to question whether you're getting away from the sheer simplicity of Minitest, and if so, whether it might be time to just use a testing framework that supports things that you want to do.

RSpec pushes things the other way and has _way too many_ features, but it's pretty easy to build conventions around the ones that you want to use. RSpec is also slightly slower to start in that it requires more files, but IMO it's unlikely that that will be the weak link in your tests.

&lt;/bikeshed&gt;

Thoughts?

**Comment by brandur** [comment]
@KensoDev Awesome man! Thanks for sending this in.

**Comment by pedro** [comment]
+1! I usually just start with minitest because it's easy to setup and I like to keep dependencies minimal at first, but sooner or later the lack of better tooling gets really annoying just like you mentioned.

Rspec is big but they did a lot of job to split it into smaller gems. Is it possible to take rspec-mocks out, for instance, considering we're using rr?

File: spec/spec_helper.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+DatabaseCleaner.strategy = :transaction
+
+ENV.update(Pliny::Utils.parse_env("#{App.root}/.env.test"))
+
+App.initialize!
+App.require!(['lib/models/**/*'])
+
+# pull in test initializers
+Pliny::Utils.require_glob("#{App.root}/test/support/**/*.rb")
```
Comment (brandur) [inline spec/spec_helper.rb]: s/test/spec/

File: vendor/pliny/lib/pliny/templates/endpoint_scaffold_acceptance_test.erb [hunk 2]
```diff
@@ -22,34 +22,34 @@ describe Endpoints::<%= plural_class_name %> do
 
   it "GET <%= url_path %>" do
     get "<%= url_path %>"
-    assert_equal 200, last_response.status
-    assert_schema_conform
+    last_response.status.should eq(200)
+    assert_schema_confirm
```
Comment (brandur) [inline vendor/pliny/lib/pliny/templates/endpoint_scaffold_acceptance_test.erb]: s/confirm/conform/ (and same thing in other specs in this suite)

File: vendor/pliny/lib/pliny/templates/endpoint_scaffold_acceptance_test.erb [hunk 3]
```diff
@@ -22,34 +22,34 @@ describe Endpoints::<%= plural_class_name %> do
...
   it "PATCH <%= url_path %>/:id" do
     header "Content-Type", "application/json"
     patch "<%= url_path %>/#{@<%= field_name %>.uuid}", MultiJson.encode({})
-    assert_equal 200, last_response.status
-    assert_schema_conform
+    last_response.status.should eq(201)
+    assert_schema_confirm
   end
 
   it "GET <%= url_path %>/:id" do
```
Comment (brandur) [inline vendor/pliny/lib/pliny/templates/endpoint_scaffold_acceptance_test.erb:50]: Doh, I must have screwed this up. Mind changing this to `DELETE`?

File: vendor/pliny/lib/pliny/templates/endpoint_scaffold_acceptance_test.erb [hunk 4]
```diff
@@ -22,34 +22,34 @@ describe Endpoints::<%= plural_class_name %> do
...
+    last_response.status.should eq(201)
+    assert_schema_confirm
   end
 =end
 
   it "GET <%= url_path %>/:id" do
     get "<%= url_path %>/#{@<%= field_name %>.uuid}"
-    assert_equal 200, last_response.status
-    assert_schema_conform
+    last_response.status.should eq(201)
```
Comment (brandur) [inline vendor/pliny/lib/pliny/templates/endpoint_scaffold_acceptance_test.erb:39]: s/201/200/ (and same for the `PATCH` and `DELETE`)

File: spec/spec_helper.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+# loaded once.
+#
+# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
+#
+ENV["RACK_ENV"] = "test"
+
+require "bundler"
+Bundler.require(:default, :test)
+
+require "minitest/autorun"
```
Comment (brandur) [inline spec/spec_helper.rb]: I think we can remove this require now.

**Comment by brandur** [comment]
@KensoDev Also, mind adding `config.mock_framework = :rr`? Thanks!

**Comment by brandur** [comment]
@pedro Awesome man, thank-you!

@KensoDev @pedro This is another somewhat arbitrary bikeshed, but like Pedro mentioned above, we technically could just use `rspec-core` and forego its other modules. For example, I'm kind of a fan of the simplicity of Minitest's assertion syntax, which isn't quite so easy to write opaquely compared to RSpec's assertions module. We could do something like `config.expect_with :stdlib`, and just use things like `assert_equal`, `assert_match`, etc. Thoughts?

**Comment by pedro** [comment]
+1! Not too concerned with rspec-expectations, but would definitely like to see mocks out.

**Comment by KensoDev** [comment]
@brandur @pedro All concerns have been addressed, I updated the PR and rebased with a single commit that describes the work. This way the git log will also make sense.

LMK if you have anything else on your mind.

**Pull request merged**